pages/sitemap/PKPSitemapHandler.inc.php

```diff
@@ -0,0 +1,152 @@
+<?php
+
+/**
+ * @file pages/sitemap/PKPSitemapHandler.inc.php
+ *
+ * Copyright (c) 2014-2018 Simon Fraser University
+ * Copyright (c) 2003-2018 John Willinsky
+ * Distributed under the GNU GPL v2. For full terms see the file docs/COPYING.
+ *
+ * @class PKPSitemapHandler
+ * @ingroup pages_sitemap
+ *
+ * @brief Produce a sitemap in XML format for submitting to search engines.
+ */
+
+import('classes.handler.Handler');
+
+define('SITEMAP_XSD_URL', 'http://www.sitemaps.org/schemas/sitemap/0.9');
+
+class PKPSitemapHandler extends Handler {
+	/**
+	 * Generate an XML sitemap for webcrawlers
+	 * Creates a sitemap index if in site context, else creates a sitemap
+	 * @param $args array
+	 * @param $request Request
+	 */
+	function index($args, $request) {
+		$context = $request->getContext();
+		if (!$context) {
+			$doc = $this->_createSitemapIndex($request);
+			header("Content-Type: application/xml");
+			header("Cache-Control: private");
+			header("Content-Disposition: inline; filename=\"sitemap_index.xml\"");
+			echo $doc->saveXml();
+		} else {
+			$doc = $this->_createContextSitemap($request);
+			header("Content-Type: application/xml");
+			header("Cache-Control: private");
+			header("Content-Disposition: inline; filename=\"sitemap.xml\"");
+			echo $doc->saveXml();
+		}
+	}
+
+	/**
+	 * Construct a sitemap index listing each context's individual sitemap
+	 * @param $request Request
+	 * @return DOMDocument
+	 */
+	function _createSitemapIndex($request) {
+		$contextDao = Application::getContextDAO();
+
+		$doc = new DOMDocument('1.0', 'utf-8');
+		$root = $doc->createElement('sitemapindex');
+		$root->setAttribute('xmlns', SITEMAP_XSD_URL);
+
+		$contexts = $contextDao->getAll(true);
+		while ($context = $contexts->next()) {
+			$sitemapUrl = $request->url($context->getPath(), 'sitemap');
+			$sitemap = $doc->createElement('sitemap');
+			$sitemap->appendChild($doc->createElement('loc', $sitemapUrl));
+			$root->appendChild($sitemap);
+		}
+
+		$doc->appendChild($root);
+		return $doc;
+	}
+
+	 /**
+	 * Construct the sitemap
+	 * @param $request Request
+	 * @return DOMDocument
+	 */
+	function _createContextSitemap($request) {
+		$context = $request->getContext();
+		$contextId = $context->getId();
+
+		$doc = new DOMDocument('1.0', 'utf-8');
+
+		$root = $doc->createElement('urlset');
+		$root->setAttribute('xmlns', SITEMAP_XSD_URL);
+
+		// Context home
+		$root->appendChild($this->_createUrlTree($doc, $request->url($context->getPath(),'index','index')));
```

Comment:
WHen I tested, I noticed that the main journal index site is coming out as:

```
http://localhost/ojs/publicknowledge/index/index
```

Google may think we mean that this is the preferred URL, when actually we want:

```
http://localhost/ojs/publicknowledge
```

Is there a way we can leave off the `index` args? Will it work with `disable_path_info` as well?

Reply:
Hmmm... No, I think without index is fine... Will do and test...